tests/e2e-tests/src/server/dot/scenarios.rs

```diff
@@ -0,0 +1,163 @@
+        ..
+    } = Graph::build(leaf_ns, Sign::No)?;
+
+    // TODO(XXX): Presently auth. hickory servers do not send referrals properly,
+    //   and so only the root NS sees queries. Once #2810 is fixed we should switch
+    //   this test to ensure the other 2 auth. nameservers see expected traffic.
+    let root_ns = nameservers.last().unwrap();
+
+    // Capture nameserver traffic on UDP 53 and TCP 853 to get both plaintext DNS
+    // and DoT requests/responses.
+    let mut tshark = Tshark::builder()
+        .filters(vec![
+            ProtocolFilter::default(),
+            ProtocolFilter::default()
+                .protocol(Protocol::Tcp)
+                .port(DOT_PORT),
+        ])
+        .ssl_keylog_file("/tmp/sslkeys.log") // See hickory.Dockerfile
+        .build(root_ns.container())?;
+
+    // Create a PDNS resolver configured for DoT probing.
+    let resolver = Resolver::new(network, root_info)
+        .custom_config(PDNS_RECURSOR_PROBE_CONFIG.to_string())
+        .start_with_subject(&Implementation::Pdns)?;
+
+    // Make a recursive query for the root FQDN using the resolver instance.
+    let client = Client::new(network)?;
+    let query_settings = *DigSettings::default().recurse();
+    let resolver_ip = resolver.ipv4_addr();
+    let resp = client.dig(query_settings, resolver_ip, RecordType::A, &FQDN::ROOT)?;
+    assert!(resp.status.is_noerror());
+
+    // Then, make a second unique query for a different name to provoke DoT probing.
+    // The pdns-recursor implementation won't do this based on the first query alone.
+    let resp = client.dig(query_settings, resolver_ip, RecordType::A, &FQDN::TEST_TLD)?;
+    assert!(resp.status.is_noerror());
+
+    // Wait until we've captured an incoming DoT probe at the nameserver.
+    tshark.wait_until(
+        |captures| {
+            captures
+                .iter()
+                .filter(|c| {
+                    matches!(c.direction, Direction::Incoming { .. }) && c.dst_port == DOT_PORT
+                })
+                .count()
+                > 1
+        },
+        Duration::from_secs(10),
+    )?;
+
+    // Now, make a final unique query for yet another name. The recursor should use DoT to satisfy
+    // the request based on its probing success.
+    let resp = client.dig(query_settings, resolver_ip, RecordType::MX, &FQDN::TEST_TLD)?;
+    assert!(resp.status.is_noerror());
+
+    // Wait until we've seen an incoming DoT query for the correct record name.
+    // This is the query from the resolver that was upgraded to DoT opportunistically.
+    let test_tld_name = FQDN::TEST_TLD.as_str().trim_end_matches('.');
+    tshark.wait_until(
+        |captures| {
+            captures.iter().any(|c| {
+                matches!(c.direction, Direction::Incoming { .. })
+                    && c.dst_port == DOT_PORT
+                    && query_name(c) == test_tld_name
+            })
+        },
+        Duration::from_secs(10),
+    )?;
```

Comment:
These second and third requests are for the same name, contrary to the comments. Also, I think the second `wait_until()` would be satisfied immediately after the first one completes, since the probe query satisfies both predicates.

Reply:
Reworked code/comments so that all queries use the same root FQDN but different record types. The second `wait_until()` checks both the name and the unique query type (MX) of the to-be-upgraded query so there's no chance it will match a probe request (these query for SOA).